### DISCUSSION TITLE
Relationship between Retype / Object.net / Bridge.net

### DISCUSSION THREAD

**Discussion by DannyBen** [retypeapp/retype#97] [General]
Hi - I am sorry for being nosey, but I am very curious about the organization side of Retype, as well as the product itself.

- On retype.com's footer, you mention object.net - so I understand Retype is a product of object.net's @geoffreymcgill.
- On https://object.net, Retype is not mentioned - why:? too new?.
- On the other hand, a product named Retyped, *is* mentioned, but....
- When clicking it, I reached Bridge.net, but....
- This bridge.net product - which is a C# to JS compiler to my understanding - is stated as "no longer maintained".
- And - also to my shallow understanding - something like bridge.net is used by Retype itself, since I saw mentions of C#, and yet I installed with `yarn add`.

So - can anyone untangle this for me?

I am mostly curious about:

- Is retype using a "no longer maintained" engine?
- Is that C# to JS compiler (bridge.net or otherwise) available to me? Sounds like a nice way to write JS (by not writing JS.... :)

Of course - since I realize this is an intrusive question, I will totally understand if it is left unanswered.

**Comment by fabriciomurta** [comment]
Well, it's not as scary as you may have thought.

> Is retype using a "no longer maintained" engine?

The only common things between bridge.net and Retype are the organization Object.NET and being written in C#. What did use Bridge was `retyped`. See the little **d** at the end of the name? Yes, a different product.

> Is that C# to JS compiler (bridge.net or otherwise) available to me? Sounds like a nice way to write JS (by not writing JS.... :)

It is available, just archived and frozen in time in [GitHub's Arctic Vault program](https://archiveprogram.github.com/arctic-vault/). There are tons of documentation about it, you can just get started from its README in the repo and wiki articles; either build it yourself or grab NuGet precompiled releases. Well, loving C# is advised. :)

**Reply by DannyBen** [reply]
Thank you (I will reply to both you and @geoffreymcgill below - as I have one more followup question).

**Comment by geoffreymcgill** [comment]
Thanks for asking. 

> On retype.com's footer, you mention object.net - so I understand Retype is a product of object.net's @geoffreymcgill.

Correct. Retype is a project/product of [Object.NET](https://object.net/). Retype will likely get spun off and become its own company at some point. Too early for that now though.

> On https://object.net, Retype is not mentioned - why:? too new?.

Just an oversight on my part as we're so focused on building Retype right now. I'll add it to my pile of TODO's.

> On the other hand, a product named Retyped, is mentioned, but....
> When clicking it, I reached Bridge.net, but....
> This bridge.net product - which is a C# to JS compiler to my understanding - is stated as "no longer maintained".

Retype**d** and Bridge.NET were related projects. Both archived now, but certainly a lot of knowledge was transferred from those projects into Retype. High performance parsing, insanely complex language parsing, processing absolutely massive amounts of text files, multi-platform runtimes, crafting buttery delicious UI's, etc, etc, are all skills honed on Bridge and Retyped that helped us eventually build **Retype**.

Retype is all new original code from Day 1. We started with a clean sheet.

Retype**d** obviously inspired the name **Retype**. Initially we were just going to repurpose the name Retyped, but in the end we were able to acquire the domain retype.com and felt **Retype** was better suited anyways, so we went with it.

> And - also to my shallow understanding - something like bridge.net is used by Retype itself, since I saw mentions of C#, and yet I installed with yarn add.

Retype does not use Bridge.NET. It likely could have, but Bridge is archived and we wanted to go in a different architectural direction with Retype. 

> Is retype using a "no longer maintained" engine?

Nope. Retype is it's own engine and it's extremely well maintained. Retype is by far the best work we've ever done, and we've done some pretty good stuff in the past.

> Is that C# to JS compiler (bridge.net or otherwise) available to me? Sounds like a nice way to write JS (by not writing JS.... :)

It's there, give it a try if you want, but the project is archived for now. Unfortunately, Bridge.NET didn't get a lot of uptake from the .NET community, so eventually we were just beating a dead horse and had to move on. Would be nice to revisit the project at some point in the future as we had a lot more to give, but for now it's archived.

Hope this helps answer your questions.

**Reply by DannyBen** [reply]
Thanks, it is a little clearer now.

So one more thing if I may - as I understand you are *still* developing in C# and transpiling it to JS right? If so - is this done with a proprietary Retype engine not available to the public? Or is it available to the public? (excluding the archived bridge.net).

**Reply by geoffreymcgill** [reply]
> So one more thing if I may - as I understand you are _still_ developing in C# and transpiling it to JS right? If so - is this done with a proprietary Retype engine not available to the public? Or is it available to the public? (excluding the archived bridge.net).

Retype is essentially two projects: 

1. Retype.App – This is a cross-platform C# app that you interact with using the [CLI](https://retype.com/guides/cli/). It does all the parsing, file reads, template creation, file output generation, etc. The app works in Mac, Win, and Linux. Only the four commands ([`watch`](https://retype.com/guides/cli/#retype-watch), [`init`](https://retype.com/guides/cli/#retype-init), [`build`](https://retype.com/guides/cli/#retype-build), and [`run`](https://retype.com/guides/cli/#retype-run)) are exposed.
2. A separate TypeScript + Vue.js + tailwindcss project – This project creates the `reytpe.js` and `retype.css` files. The project is also responsible for maintaining the HTML templates used by Retype.App.

Check out the [credits](https://retype.com/about/#credits) for a list of all the libraries we are currently using. 

No C# to JS transpiling is done. We use TypeScript to generate the `retype.js` file.

Eventually, we might replace Vue JS. We've slowly be replacing misc Vue pieces, but there is a bigger re-write coming at some point. That won't have any effect on your website though. Things will just magically get better, as they do with every release.

There's also a monster big unit test project. All those unit tests are automatically run by the CI on every commit to help ensure everything is running smooth.

Hope this helps.

**Reply by DannyBen** [reply]
Thanks, clearer now.
